Bulk Enrollment

azorica opened the pull request:
Bulk Enrollment for each User ID and selected Program ID:

- added checkboxes to users dataTable for selecting users 
- added button "Enroll users to program"
- added enrollment Dialog modal for selecting the enrollment program
- each selected user enrolls to selected program by clicking "Enroll" button in enrollment Dialog

azorica committed:
Bulk Enrollment for User IDs and Program ID
Bulk Enrollment for each User ID and selected Program ID

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+          <v-autocomplete
+            v-model="programId"
+            :items="listPrograms"
+            :loading="isLoading"
```
What does this prop do? Actually you don't need it, because you already have list of program on this page so there is no need for fetching anything form the server.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+const el = vm => vm.$children[0].$refs.dialog;
+
+export default {
+  name: 'enrollment-dialog',
```
Component should have some other name because we already have enrollment-dialog component. Maybe bulk-enrollment-dialog? And it doesn't need to have it's own folder.
Change it in the parent component also, I didn't want to write it on multiple places.

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -3,6 +3,9 @@
     <v-flex class="mt-5">
       <v-toolbar color="#f5f5f5" flat>
         <v-spacer/>
+        <v-btn @click.stop="enrollUsersDialog()" color="success" outline>
```
Shouldn't this button be disabled/hidden if there is no users selected?

Also consider using activator slot inside v-dialog for triggering it, like other enrollment-dialog component do.

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -3,6 +3,9 @@
       <v-toolbar color="#f5f5f5" flat>
         <v-spacer/>
+        <v-btn @click.stop="enrollUsersDialog()" color="success" outline>
+          Enroll To Program
```
User can be enrolled only in program so you can omit 'To Program'

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+  },
+  methods: {
+    bulkEnroll() {
+      let thisState = this;
```
You don't need to store 'this' in another variable, use arrow function in forEach so you will have the right 'this'.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+  methods: {
+    bulkEnroll() {
+      let thisState = this;
+      forEach(this.userIds, function (userId) {
```
Create separate API endpoint for handling bulk import, beacuse of performance reasons and better error handling.

Another note if you have this kind of situation (list of async function) use something like Promise.all/map/each so you can wait for all promises to finish.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+        });
+      });
+      this.close();
+      this.$emit('enrolled');
```
You emit enrolled event but you don't do anything in the parent component with it (and you should 🤡 )

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+        enrollmentApi.create(pick(thisState, ['studentId', 'programId'])).then(() => {
+        });
+      });
+      this.close();
```
You should close dialog after all API calls finish.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+    bulkEnroll() {
+      let thisState = this;
+      forEach(this.userIds, function (userId) {
+        thisState.studentId = userId;
```
You don't need to bind userId variable to thisState (or this)

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+      let thisState = this;
+      forEach(this.userIds, function (userId) {
+        thisState.studentId = userId;
+        enrollmentApi.create(pick(thisState, ['studentId', 'programId'])).then(() => {
```
If you did the line above so you can with pick create an object suited for this API call it would be better to create an object without pick.
If you want to use shorthands, you can call iteratee in the forEach studentId instead of userId, and destructure programId from this object.

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -25,6 +28,12 @@
           :must-sort="true">
           <template slot="items" slot-scope="{ item }">
+            <td class="select">
+              <input
```
Use v-checkbox (vuetify component).

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -67,16 +80,20 @@ export default {
       totalItems: 0,
-      isLoading: false
+      isLoading: false,
+      checkedItems: []
```
Why you have both checkedItems and checkedUsers? You can use only one variable.

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -113,12 +134,28 @@ export default {
+  thead {
+    th.select {
+      padding: 0;
+      text-align: center;
```
You can use vuetify text alignement helpers instead of writing it yourself.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+            @blur="focusTrap.unpause()"
+            label="Program"
+            placeholder="Start typing to Search"
+            prepend-icon="search"
```
You should use mdi webfont instead of material.io/icons

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+  data() {
+    return {
+      programId: null,
+      userId: null,
```
You don't need this variable here.

mandryllo commented on client/admin/components/users/Enrollments/EnrollmentDialog.vue:
```diff
@@ -0,0 +1,84 @@
+      }
+    },
+    ...mapState('programs', { programs: 'items' }),
+    listPrograms() {
```
programList?
Maybe call this variable programs, and in the line above renamed it to something else.

azorica committed:
EnrollmentDialog renamed  to BulkEnrollmentDialog

azorica committed:
Various repairs to bulk enrollment

Missing refactor to BulkEnroll method

azorica committed:
Enable/Disable Enroll Button

Enable/Disable Enroll Button depending on selected uesr roe checkboxes

azorica committed:
Enable/Disable Enroll Button - minor update

azorica committed:
Disabling Enroll button with computed property

azorica committed:
User Bulk enrollment to program with Promise.all

azorica committed:
Bulk create with Promise.each

azorica committed:
Response messages for bulk enroll

Enrollment controller response messages for each enroll with user ID and Program ID

azorica committed:
Bulk and Single Crate nad redirect to Program

Seperated enrollment creation for single user and multiple users in enrollment controller and added redirection to program page after bulk enrollment.

azorica committed:
Users data table - select all enabled

selecting all Users enabled for bulk enrollment

azorica committed:
User data table checkboxes alignment and styling

azorica committed:
Enrollment bulkEnroll API endpoint added

User data table checkbox style cleanup

azorica committed:
Bulk Enrollment v-alert messages in dialog

Displaying the message for every enroll attempt in bulk enroll

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+</template>
+
+<script>
+import enrollmentApi from '@/admin/api/enrollment';
```
you can rename this to just api

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+            placeholder="Start typing to Search"
+            prepend-icon="mdi-magnify"
+            clearable
+            name="program"/>
```
reorder props by importance

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+  name: 'bulk-enrollment-dialog',
+  mixins: [withValidation(), withFocusTrap({ el })],
+  props: {
+    enrollDisabled: { type: Boolean, default: true },
```
rename this prop to just disabled

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+      visible: false,
+      programId: null,
+      enrollBtnDisabled: false,
+      closeBtnDisabled: false,
```
enrollBtnDisabled and closeBtnDisabled will
always have thesame state so you can use only one of them, maybe enrollDisabled

and this should be computed property which will check if program is selected

azorica replied:
They won't always have the same state: when the enroll process finishes, enroll button is disabled to prevent new enroll but the close button is enabled for exiting the dialog.

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+      programId: null,
+      enrollBtnDisabled: false,
+      closeBtnDisabled: false,
+      enrollBtnLoading: false,
```
rename it to something like enrolling or enrollInProgress

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+          this.enrollmentMessages = res.messages;
+          res.messages.length ? this.showAlerts = true : this.showAlerts = true;
+          this.enrollFinished();
+        });
```
catch errors to set flags correctly in case of any server errors

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+          <v-btn :disabled="closeBtnDisabled" @click="close">Cancel</v-btn>
+          <v-btn :disabled="enrollBtnDisabled" :loading="enrollBtnLoading" color="success" type="submit">Enroll</v-btn>
+        </v-card-actions>
+        <v-alert
```
there is a better way of dislaying errors, maybe error-messages prop on v-autocomplete

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -19,21 +22,33 @@
-            </td>
+          :pagination.sync="dataTable"
+          select-all
+          item-key="id"
```
this is set by default so you can omit it

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -113,12 +133,17 @@ export default {
 };
 </script>
 
-<style lang="scss" scoped>
```
use scoped attribute or this css will be applied on all components

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -72,18 +96,14 @@ export default {
-      { text: 'Date Created', value: 'createdAt' },
-      { text: 'Actions', value: 'email', sortable: false }
-    ]),
+    disableEnroll() {
```
you can omit this property and do this check in the template

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+      this.visible = false;
+      this.enrollDialogDefault();
+    },
+    enrollDialogDefault() {
```
you can do this without so many methods for setting flags

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+  methods: {
+    bulkEnroll() {
+      this.enrollPending();
+      enrollmentApi.bulkEnroll({ users: this.users, programId: this.programId })
```
return api call

destructure users and programId from this and use shorthands for api call arguments

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+
+export default {
+  name: 'bulk-enrollment-dialog',
+  mixins: [withValidation(), withFocusTrap({ el })],
```
you add validation but you dont use it

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +37,33 @@ function create({ body }, res) {
+  const { users, programId } = body;
+  const enrollMessages = [];
+  return Promise.each(users, user => {
+    console.log('bbb... ' + user);
```
please dont commit console messages

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +37,33 @@ function create({ body }, res) {
+  const enrollMessages = [];
+  return Promise.each(users, user => {
+    console.log('bbb... ' + user);
+    let data = { studentId: user.id, programId: programId };
```
why let?
you can use shorthand for programId
you can even destructure id(and rename it to studentId) from user in Promise.each to write this cleaner

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +37,33 @@ function create({ body }, res) {
 }
 
+function bulkEnroll({ body }, res) {
+  const { users, programId } = body;
```
ypu can destructure this on the line above

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +37,33 @@ function create({ body }, res) {
+  return Promise.each(users, user => {
+    console.log('bbb... ' + user);
+    let data = { studentId: user.id, programId: programId };
+    return Enrollment.findOne({ where: data, paranoid: false })
```
it would be better to find all enrollments for this combo of users and program before Promise.each and then in each promise check if enrollment for that user exist

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -2,9 +2,11 @@
 
 const { Enrollment, Sequelize } = require('../common/database');
 const { createError } = require('../common/errors');
+const { bulkEnrollmentMessage } = require('../common/messages');
```
order alphabetical

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +37,33 @@ function create({ body }, res) {
 
+function bulkEnroll({ body }, res) {
+  const { users, programId } = body;
+  const enrollMessages = [];
```
please omit enroll messages, you dont need report for every user what happened

if you want you can just return number of already enrolled users

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,116 @@
+        <v-card-actions>
+          <v-spacer/>
+          <v-btn :disabled="closeBtnDisabled" @click="close">Cancel</v-btn>
+          <v-btn :disabled="enrollBtnDisabled" :loading="enrollBtnLoading" color="success" type="submit">Enroll</v-btn>
```
this is more then 80 characters in one line

azorica committed:
Bulk enroll cleanup

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+            :disabled="enrollDisabled"
+            @click="close">
+            Cancel
+          </v-btn>
```
you can fit cancel button in one line

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -113,12 +132,17 @@ export default {
   background-color: #fff;
 }
+
+.v-input--checkbox {
```
use scoped attribute, select any element in your component(by giving v-data-table a class or selecting some of existing classes) and than use deep scoping on .v-input--checkbox

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+    return {
+      visible: false,
+      programId: null,
+      enrollDisabled: false,
```
this should be a computed property which will check if a program is selected and also if enrollInProgress is set, and than you don't need enrollDialogDefault, enrollPending and enrollFinished methods because you just need to set enrollInProgress and enrollDisabled will update because of reactivity

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+    bulkEnroll() {
+      this.enrollPending();
+      return api.bulkEnroll({ users: this.users, programId: this.programId })
+        .catch(() => (this.enrollmentMessage = 'Error! Unable to enroll Users!'))
```
in case of server errors, you should also set enrollInProgress to false

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+};
+</script>
+
+<style lang="scss">
```
always use scoped on your components, if it doesn't work with scoped than you are not selecting elements correctly

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+            Enroll
+          </v-btn>
+        </v-card-actions>
+        <v-card-text>{{ enrollmentMessage }}</v-card-text>
```
please try out :error-messages prop on v-autocomplete to see how that looks

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+  methods: {
+    bulkEnroll() {
+      this.enrollPending();
+      return api.bulkEnroll({ users: this.users, programId: this.programId })
```
destructure this
```js
const { users, programId } = this;
return api.bulkEnroll({ users, programId })...
```

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+      enrollDisabled: false,
+      enrollInProgress: false,
+      enrollmentMessage: '',
+      showAlert: false
```
you don't use this prop anywhere

mandryllo commented on client/admin/components/users/BulkEnrollmentDialog.vue:
```diff
@@ -0,0 +1,129 @@
+      return api.bulkEnroll({ users: this.users, programId: this.programId })
+        .catch(() => (this.enrollmentMessage = 'Error! Unable to enroll Users!'))
+        .then(res => {
+          if (res.message.type === 'success') {
```
you can write this error handling nicer
```
set enrollInProgress to false
if (error) return (set errors)
this.close()
```

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -19,21 +22,32 @@
+          select-all>
+          <template slot="items" slot-scope="props">
+            <tr>
+              <td class="text-xs-center">
```
omit this class here, because it doesn't do anything

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -19,21 +22,32 @@
+                  v-model="props.selected"
+                  primary
+                  hide-details
+                ></v-checkbox>
```
self-close v-checkbox component because you dont pass any content to it
than you can fit whole this td element in one line

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -64,6 +79,14 @@ export default {
   name: 'user-list',
   data() {
     return {
+      headers: [
```
headers should be computed prop

azorica committed:
Bulk enrollment code cleanup

mandryllo commented on client/admin/components/users/index.vue:
```diff
@@ -3,6 +3,9 @@
         <v-spacer/>
+        <bulk-enrollment-dialog
+          :disabled="disableEnroll"
+          :users="checkedItems"/>
```
this can be in just one line

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -3,8 +3,11 @@
 const { createError } = require('../common/errors');
 const HttpStatus = require('http-status');
+const includes = require('lodash/includes');
+const find = require('lodash/find');
```
sort in alphabetical order

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -3,8 +3,11 @@
 const { Enrollment, Sequelize } = require('../common/database');
 const { createError } = require('../common/errors');
 const HttpStatus = require('http-status');
+const includes = require('lodash/includes');
```
you dont use includes anywhere

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +38,42 @@ function create({ body }, res) {
 }
 
+function bulkEnroll({ body: { users, programId } }, res) {
+  const userIds = map(users, 'id');
```
do this on the frontend so you dont send whole user object through api

azorica replied:
the whole user Obj will be needed in the future for error handling

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +38,42 @@ function create({ body }, res) {
 
+function bulkEnroll({ body: { users, programId } }, res) {
+  const userIds = map(users, 'id');
+  return Enrollment.findAll({
```
change bulkEnroll to async function and use await here to avoid promise nesting and to have cleaner code

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +38,42 @@ function create({ body }, res) {
+    paranoid: false
+  })
+  .then(existingUsers => {
+    let enrollMessage = { type: 'success', text: '' };
```
you can return error count without enrollMessage

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +38,42 @@ function create({ body }, res) {
+    where: {
+      studentId: userIds,
+      programId
+    },
```
you can fit the whole where object in one line or create separate 'where' object before you use it in findAll

mandryllo commented on server/enrollment/enrollment.controller.js:
```diff
@@ -35,6 +38,42 @@ function create({ body }, res) {
+      if (existingUser.deletedAt) {
+        existingUser.setDataValue('deletedAt', null);
+        return existingUser.save();
+      } else {
```
you dont need else here when you have return inside both above if-s

azorica committed:
Cleanup and different error handling

azorica committed:
Async/await promise syntax in bulkEnroll

azorica committed:
Cleanup 💄

azorica committed:
Merge branch 'develop' into feature/users-bulk-enrollement

# Conflicts:
#	client/admin/components/users/index.vue

vladimyr committed:
Use same route for single & bulk enrollment :tada:

- determining enrollment type by payload introspection:
  Number vs Array(Number)
- refactoring bulk enrollment dialog

vladimyr committed:
Cleanup :lipstick:

underscope committed:
Merge pull request #55 from ExtensionEngine/feature/refactor-bulk-enrollment

Use same route for single & bulk enrollment

underscope committed:
Removed unused route :wrench:

underscope committed:
Removed unused var

underscope committed:
Cleanup

- Omit defaults
- Improved naming
- Force sorting (reverted)

underscope committed:
Cleanup :lipstick:

- Omit blank
- Extra elevation (already on parent)

vladimyr committed:
Connect BulkEnrollmentDialog with store and http client :wrench:
